From b1ecfc228399973e3d138a14cf1c799c5dc2f8a0 Mon Sep 17 00:00:00 2001
From: Pavel Raiskup <praiskup@redhat.com>
Date: Mon, 19 Nov 2012 14:08:09 +0100
Subject: [PATCH] fix crash when file is specified multiple times

---
 src/names.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/names.c b/src/names.c
index 34ebac2..c921164 100644
--- a/src/names.c
+++ b/src/names.c
@@ -871,7 +871,7 @@ void
 collect_and_sort_names (void)
 {
   struct name *name;
-  struct name *next_name, *prev_name;
+  struct name *next_name, *prev_name = NULL;
   int num_names;
   struct stat statbuf;
   Hash_table *nametab;
@@ -947,6 +947,7 @@ collect_and_sort_names (void)
 		{
 		  if (p->child)
 		    rebase_child_list (p->child, name);
+		  hash_delete (nametab, name);
 		  /* FIXME: remove_directory (p->caname); ? */
 		  remname (p);
 		  free_name (p);
-- 
1.7.11.7

